inference-engine/src/inference_engine/compilation_context.cpp

```diff
@@ -0,0 +1,209 @@
+    std::streampos m_pos = {};
+
+public:
+    std::size_t getResult() const { return m_res; }
+    std::streamsize xsputn(const char* s, std::streamsize n) override {
+        std::string str(s, n);
+        m_res = hash_combine(m_res, str);
+        m_pos += n;
+        return n;
+    }
+};
+
+//////////////////////////////////////////////////
+
+std::string NetworkCompilationContext::calculateFileInfo(const std::string& filePath) {
+    std::string res;
+    struct stat result;
+    size_t seed {};
+    seed = hash_combine(seed, filePath);
+    if (stat(filePath.c_str(), &result) == 0) {
+        seed = hash_combine(seed, result.st_mtime);
+        seed = hash_combine(seed, result.st_size);
+    }
+    return std::to_string(seed);
+}
+
+std::string NetworkCompilationContext::computeHash(const CNNNetwork& network,
+                               const std::map<std::string, std::string>& compileOptions) {
+    OV_ITT_SCOPED_TASK(itt::domains::IE_LT, "NetworkCompilationContext::computeHash - CNN");
+    OstreamHashWrapper xmlHash;
+    OstreamHashWrapper binHash;
+    std::ostream xml(&xmlHash);
+    std::ostream bin(&binHash);
+
+    IE_ASSERT(network.getFunction());
+
+    // 1. Serialize
+    CNNNetwork net(network);
+    ngraph::pass::Serialize serializer(xml, bin,
+        ngraph::pass::Serialize::Version::IR_V10);
+    serializer.run_on_function(net.getFunction());
+
+    // 2. Compute hash on serialized data and options
+    size_t seed {};
+    seed = hash_combine(seed, xmlHash.getResult());
+    seed = hash_combine(seed, binHash.getResult());
+    for (const auto& kvp : compileOptions) {
+        seed = hash_combine(seed, kvp.first + kvp.second);
+    }
+
+    // 3. Add runtime information which was not serialized
+    for (const auto& op : network.getFunction()->get_ordered_ops()) {
+        ngraph::Node::RTMap rt = op->get_rt_info();
+
+        auto affinity_it = rt.find("affinity");
+        seed = hash_combine(seed, op->get_friendly_name());
+
+        if (rt.end() != affinity_it) {
+            auto affinity = std::dynamic_pointer_cast<ngraph::VariantWrapper<std::string>>(affinity_it->second);
+            seed = hash_combine(seed, std::string(affinity->get()));
+        }
+
+        auto priorities_it = rt.find("PrimitivesPriority");
+        if (rt.end() != priorities_it) {
+            auto primPriority = std::dynamic_pointer_cast<ngraph::VariantWrapper<std::string>>(priorities_it->second);
+            seed = hash_combine(seed, std::string(primPriority->get()));
+        }
+
+        if (const auto& c = std::dynamic_pointer_cast<ngraph::opset6::Constant>(op)) {
+            auto data = reinterpret_cast<const std::uint8_t *>(c->get_data_ptr());
+            auto data_size = c->get_element_type().size() * ngraph::shape_size(c->get_shape());
+
+            std::uint64_t sum = 0;
+            for (size_t i = 0; i < data_size; ++i) {
+                sum += data[i];
+            }
+            seed = hash_combine(seed, sum);
```

Comment:
looks like hash for bin data is taken in the line https://github.com/openvinotoolkit/openvino/pull/4661/files/14a495da98e912450e89f2d72705698f0812c3c0..00f875ac1801d93264e11b4b25c80e86f0b93ee3#diff-20757477b641f38e97a3c5c028c6b5d5a3c943c53f75b432211f5e47f274b814R53

BTW, have you compare performance? E.g.:
1. Before: weights hash is calculated as hash of sum of weights represented as bytes.
2. After: weights hash is computed directly on each byte of weights represented as string

Looks like "after" we need more compute time..